docs/ql-style-guide.md

```diff
@@ -0,0 +1,434 @@
+1. Declarations in query files *should* be documented.
+1. Use `/** ... */` for documentation, even for single line comments.
+1. For single-line documentation, the `/**` and `*/` are written on the same line as the comment.
+1. For multi-line documentation, the `/**` and `*/` are written on separate lines. There is a `*` preceding each comment line, aligned on the first `*`.
+1. Library files (`.qll` files) *should* be documented with QLDoc.
+1. Query files, except for tests, *must* have a QLDoc query documentation comment.
+1. Documentation *should* be appropriate for users of the library. Put documentation for maintainers in normal comments.
+1. Predicates that do not have a result *should* be documented `/** Holds if ... */`
+1. Predicates that have a result *should* be documented `/** Gets ... */`
+1. All predicate parameters *should* be referred to in the predicate documentation.
+1. Reference names, such as types and parameters, using backticks `` ` ``.
+1. Give examples of code in the target language, enclosed in ```` ``` ````  or `` ` ``.
+1. Classes *should* be documented in the singular, for example `/* An expression. */`
+1. Where a class denotes a generic concept with subclasses, list those subclasses.
+1. Declarations that are deprecated *should* be documented as `DEPRECATED: ...`
+1. Declarations that are for internal use *should* be documented as `INTERNAL: Do not use`.
+1. Use full sentences, with capital letters and full stops.
+
+
+Examples:
+
+```ql
+/** Provides logic for determining constant expressions. */
+```	
+
+```ql
+/**
+ * Holds if the qualifier of this call has type `qualifierType`.
+ * `isExactType` indicates whether the type is exact, that is, whether
+ * the qualifier is guaranteed not to be a subtype of `qualifierType`.
+ */
+```	
+```ql
+/**
+ * A delegate declaration, for example
+ * ```
+ * delegate void Logger(string text); 
+ * ```
+ */
+class Delegate extends ...
+```
+
+```ql
+/**
+ * An element that can be called.
+ *
+ * Either a method (`Method`), a constructor (`Constructor`), a destructor
+ * (`Destructor`), an operator (`Operator`), an accessor (`Accessor`),
+ * an anonymous function (`AnonymousFunctionExpr`), or a local function
+ * (`LocalFunction`).
+ */
+class Callable extends ...
+```
+
+```ql
+/** DEPRECATED: Use `getAnExpr()` instead. */
+deprecated Expr getInitializer()
+```
+
+```ql
+/**
+ * INTERNAL: Do not use.
+ */
+```
+
+## Formulas
+1. *Prefer* one *conjunct* per line.
+1. Write the `and` at the end of the line. This also applies in `where` clauses.
+1. *Prefer* to write the `or` keyword on its own line.
+1. The `or` keyword *may* be written at the end of a line, or within a line, provided that it has no unparenthesised `and` operands.
+1. Single-line formulas *may* be used in order to save space or add clarity, particularly in the *body* of a *quantifier/aggregation*. 
+1. *Avoid* additional brackets to clarify the precedence of `not`, `and` and `or`.

```

Comment:
Isn't this the exact opposite of previous policy re: `and` / `or`?

Reply:
I'm just going to drop this item for now and perhaps reinstate it when we decide whether we want to write `(A and B) or C` or `A and B or C`